File: silencendo/TMplatform/src/sources/manager.ts
import { KnowledgeSource, SourceState } from './types';
import * as fs from 'fs';
import * as path from 'path';
import { randomUUID } from 'crypto';

const SOURCES_FILE = path.join('.bot', 'sources.json');

export class SourceManager {
  private state: SourceState;

  constructor() {
    this.state = {
      allSources: [],
      activeSourceIds: [],
      nextSourceNumber: 1,
    };
    this.loadSources();
  }

  private ensureBotDir() {
    if (!fs.existsSync('.bot')) {
      fs.mkdirSync('.bot', { recursive: true });
    }
  }

  private loadSources(): void {
    this.ensureBotDir();
    
    if (fs.existsSync(SOURCES_FILE)) {
      try {
        const data = JSON.parse(fs.readFileSync(SOURCES_FILE, 'utf8'));
        // Convert date strings back to Date objects
        const loadedSources = data.allSources.map((source: any) => ({
          ...source,
          createdAt: new Date(source.createdAt)
        }));
        
        // Calculate nextSourceNumber based on highest number in loaded sources
        const maxNumber = loadedSources.length > 0
          ? Math.max(...loadedSources.map((s: any) => s.number))
          : 0;
        const nextNumber = Math.max(maxNumber + 1, data.nextSourceNumber || 1);
        
        this.state = {
          ...data,
          allSources: loadedSources,
          nextSourceNumber: nextNumber
        };
      } catch (error) {
        console.error('Error loading sources:', error);
        // Use default state if loading fails
      }
    } else {
      // Initialize with empty state
      this.saveSources();
    }
  }

  private saveSources(): void {
    this.ensureBotDir();
    // Convert Date objects to strings for JSON serialization
    const serializableState = {
      ...this.state,
      allSources: this.state.allSources.map(source => ({
        ...source,
        createdAt: source.createdAt.toISOString()
      })),
      nextSourceNumber: this.state.nextSourceNumber
    };
    fs.writeFileSync(SOURCES_FILE, JSON.stringify(serializableState, null, 2));
  }

  public addSource(type: 'file' | 'url', path: string): KnowledgeSource {
    // Check if source already exists
    const existingSource = this.state.allSources.find(s => s.path === path);
    if (existingSource) {
      return existingSource;
    }

    // Find the lowest available number
    let availableNumber = 1;
    const usedNumbers = new Set(this.state.allSources.map(s => s.number));
    while (usedNumbers.has(availableNumber)) {
      availableNumber++;
    }
    
    const newSource: KnowledgeSource = {
      id: randomUUID(),
      number: availableNumber,
      type,
      path,
      title: path.split('/').pop() || path,
      createdAt: new Date(),
      metadata: {}
    };
    
    // Update nextSourceNumber to be the next available number
    this.state.nextSourceNumber = Math.max(this.state.nextSourceNumber, availableNumber + 1);

    this.state.allSources.push(newSource);
    this.saveSources();
    return newSource;
  }

  public listSources(): KnowledgeSource[] {
    return this.state.allSources;
  }

  public setSourceActive(sourceId: string, active: boolean): void {
    if (active && !this.state.activeSourceIds.includes(sourceId)) {
      this.state.activeSourceIds.push(sourceId);
    } else if (!active) {
      this.state.activeSourceIds = this.state.activeSourceIds.filter(id => id !== sourceId);
    }
    this.saveSources();
  }

  public setActiveSources(sourceIds: string[]): void {
    this.state.activeSourceIds = sourceIds;
    this.saveSources();
  }

  public clearActiveSources(): void {
    this.state.activeSourceIds = [];
    this.saveSources();
  }

  public getActiveSources(): KnowledgeSource[] {
    return this.state.allSources.filter(source => 
      this.state.activeSourceIds.includes(source.id)
    );
  }

  public getSourceById(id: string): KnowledgeSource | undefined {
    // First try exact UUID match
    const exactMatch = this.state.allSources.find(source => source.id === id);
    if (exactMatch) {
      return exactMatch;
    }
    
    // Then try partial UUID match (for backward compatibility with 4-char prefixes)
    const partialMatch = this.state.allSources.find(source => source.id.startsWith(id));
    if (partialMatch) {
      return partialMatch;
    }
    
    // Then try sequential number match
    const number = parseInt(id, 10);
    if (!isNaN(number)) {
      const numberMatch = this.state.allSources.find(source => source.number === number);
      if (numberMatch) {
        return numberMatch;
      }
    }
    
    return undefined;
  }
  
  public getSourceByNumber(number: number): KnowledgeSource | undefined {
    return this.state.allSources.find(source => source.number === number);
  }
  
  public getSourceIdByNumber(number: number): string | undefined {
    const source = this.state.allSources.find(source => source.number === number);
    return source?.id;
  }

  public getAllSources(): KnowledgeSource[] {
    return this.state.allSources;
  }

  public updateSourceContent(sourceId: string, newContent: string): boolean {
    const source = this.state.allSources.find(s => s.id === sourceId);
    if (!source || source.type !== 'file') {
      return false;
    }

    try {
      // Write the new content back to the file
      fs.writeFileSync(source.path, newContent, 'utf8');
      
      // After updating the file, we should re-ingest it to update the chunks
      return true;
    } catch (error) {
      console.error('Error updating source content:', error);
      return false;
    }
  }

  public getSourceContent(sourceId: string): string | null {
    const source = this.state.allSources.find(s => s.id === sourceId);
    if (!source || source.type !== 'file') {
      return null;
    }

    try {
      return fs.readFileSync(source.path, 'utf8');
    } catch (error) {
      console.error('Error reading source content:', error);
      return null;
    }
  }

  public removeSource(sourceId: string): boolean {
    // Try to find the source by UUID, partial UUID, or sequential number
    let matchingSource: KnowledgeSource | undefined = undefined;
    
    // First try exact UUID match
    const exactMatch = this.state.allSources.find(s => s.id === sourceId);
    if (exactMatch) {
      matchingSource = exactMatch;
    }
    
    // Then try partial UUID match
    if (!matchingSource) {
      const partialMatch = this.state.allSources.find(s => s.id.startsWith(sourceId));
      if (partialMatch) {
        matchingSource = partialMatch;
      }
    }
    
    // Then try sequential number match
    if (!matchingSource) {
      const number = parseInt(sourceId, 10);
      if (!isNaN(number)) {
        const numberMatch = this.state.allSources.find(s => s.number === number);
        if (numberMatch) {
          matchingSource = numberMatch;
        }
      }
    }
    
    if (!matchingSource) {
      return false;
    }
    
    const initialLength = this.state.allSources.length;
    this.state.allSources = this.state.allSources.filter(source => source.id !== matchingSource!.id);
    
    // Also remove from active sources if it was active
    this.state.activeSourceIds = this.state.activeSourceIds.filter(id => id !== matchingSource!.id);
    
    if (this.state.allSources.length < initialLength) {
      this.saveSources();
      return true;
    }
    return false;
  }
}

File: silencendo/TMplatform/sources/manager.go
package sources

import (
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"io/ioutil"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

const SOURCES_FILE = ".bot/sources.json"

type SourceManager struct {
	state SourceState
}

func NewSourceManager() *SourceManager {
	sm := &SourceManager{
		state: SourceState{
			AllSources:       []KnowledgeSource{},
			ActiveSourceIds:  []string{},
			NextSourceNumber: 1,
		},
	}
	sm.loadSources()
	return sm
}

func (sm *SourceManager) ensureBotDir() {
	if _, err := os.Stat(".bot"); os.IsNotExist(err) {
		os.MkdirAll(".bot", 0755)
	}
}

func (sm *SourceManager) loadSources() {
	sm.ensureBotDir()

	if _, err := os.Stat(SOURCES_FILE); err == nil {
		data, err := ioutil.ReadFile(SOURCES_FILE)
		if err == nil {
			var tempState struct {
				AllSources []struct {
					ID        string            `json:"id"`
					Number    int               `json:"number"`
					Type      string            `json:"type"`
					Path      string            `json:"path"`
					Title     string            `json:"title"`
					CreatedAt string            `json:"createdAt"`
					Metadata  map[string]string `json:"metadata,omitempty"`
				} `json:"allSources"`
				ActiveSourceIds  []string `json:"activeSourceIds"`
				NextSourceNumber int      `json:"nextSourceNumber"`
			}

			err = json.Unmarshal(data, &tempState)
			if err == nil {
				// Convert date strings back to time.Time objects
				var loadedSources []KnowledgeSource
				for _, source := range tempState.AllSources {
					createdAt, _ := time.Parse(time.RFC3339, source.CreatedAt)
					loadedSources = append(loadedSources, KnowledgeSource{
						ID:        source.ID,
						Number:    source.Number,
						Type:      source.Type,
						Path:      source.Path,
						Title:     source.Title,
						CreatedAt: createdAt,
						Metadata:  source.Metadata,
					})
				}

				// Calculate nextSourceNumber based on highest number in loaded sources
				maxNumber := 0
				for _, s := range loadedSources {
					if s.Number > maxNumber {
						maxNumber = s.Number
					}
				}
				nextNumber := maxNumber + 1
				if nextNumber < tempState.NextSourceNumber {
					nextNumber = tempState.NextSourceNumber
				}

				sm.state = SourceState{
					AllSources:       loadedSources,
					ActiveSourceIds:  tempState.ActiveSourceIds,
					NextSourceNumber: nextNumber,
				}
			} else {
				// Use default state if loading fails
				sm.saveSources()
			}
		} else {
			// Use default state if loading fails
			sm.saveSources()
		}
	} else {
		// Initialize with empty state
		sm.saveSources()
	}
}

func (sm *SourceManager) saveSources() {
	sm.ensureBotDir()

	// Convert time.Time objects to strings for JSON serialization
	var serializableSources []struct {
		ID        string            `json:"id"`
		Number    int               `json:"number"`
		Type      string            `json:"type"`
		Path      string            `json:"path"`
		Title     string            `json:"title"`
		CreatedAt string            `json:"createdAt"`
		Metadata  map[string]string `json:"metadata,omitempty"`
	}

	for _, source := range sm.state.AllSources {
		serializableSources = append(serializableSources, struct {
			ID        string            `json:"id"`
			Number    int               `json:"number"`
			Type      string            `json:"type"`
			Path      string            `json:"path"`
			Title     string            `json:"title"`
			CreatedAt string            `json:"createdAt"`
			Metadata  map[string]string `json:"metadata,omitempty"`
		}{
			ID:        source.ID,
			Number:    source.Number,
			Type:      source.Type,
			Path:      source.Path,
			Title:     source.Title,
			CreatedAt: source.CreatedAt.Format(time.RFC3339),
			Metadata:  source.Metadata,
		})
	}

	tempState := struct {
		AllSources       interface{} `json:"allSources"`
		ActiveSourceIds  []string    `json:"activeSourceIds"`
		NextSourceNumber int         `json:"nextSourceNumber"`
	}{
		AllSources:       serializableSources,
		ActiveSourceIds:  sm.state.ActiveSourceIds,
		NextSourceNumber: sm.state.NextSourceNumber,
	}

	data, err := json.MarshalIndent(tempState, "", "  ")
	if err == nil {
		ioutil.WriteFile(SOURCES_FILE, data, 0644)
	}
}

func (sm *SourceManager) AddSource(sourceType string, path string) *KnowledgeSource {
	// Check if source already exists
	for _, source := range sm.state.AllSources {
		if source.Path == path {
			return &source
		}
	}

	// Find the lowest available number
	usedNumbers := make(map[int]bool)
	for _, source := range sm.state.AllSources {
		usedNumbers[source.Number] = true
	}

	availableNumber := 1
	for usedNumbers[availableNumber] {
		availableNumber++
	}

	id := generateUUID()

	newSource := KnowledgeSource{
		ID:        id,
		Number:    availableNumber,
		Type:      sourceType,
		Path:      path,
		Title:     filepath.Base(path),
		CreatedAt: time.Now(),
		Metadata:  make(map[string]string),
	}

	// Update nextSourceNumber to be the next available number
	if sm.state.NextSourceNumber <= availableNumber {
		sm.state.NextSourceNumber = availableNumber + 1
	}

	sm.state.AllSources = append(sm.state.AllSources, newSource)
	sm.saveSources()
	return &newSource
}

func (sm *SourceManager) ListSources() []KnowledgeSource {
	return sm.state.AllSources
}

func (sm *SourceManager) SetSourceActive(sourceId string, active bool) {
	if active {
		found := false
		for _, id := range sm.state.ActiveSourceIds {
			if id == sourceId {
				found = true
				break
			}
		}
		if !found {
			sm.state.ActiveSourceIds = append(sm.state.ActiveSourceIds, sourceId)
		}
	} else {
		var newActiveIds []string
		for _, id := range sm.state.ActiveSourceIds {
			if id != sourceId {
				newActiveIds = append(newActiveIds, id)
			}
		}
		sm.state.ActiveSourceIds = newActiveIds
	}
	sm.saveSources()
}

func (sm *SourceManager) SetActiveSources(sourceIds []string) {
	sm.state.ActiveSourceIds = sourceIds
	sm.saveSources()
}

func (sm *SourceManager) ClearActiveSources() {
	sm.state.ActiveSourceIds = []string{}
	sm.saveSources()
}

func (sm *SourceManager) GetActiveSources() []KnowledgeSource {
	var activeSources []KnowledgeSource
	for _, source := range sm.state.AllSources {
		for _, id := range sm.state.ActiveSourceIds {
			if source.ID == id {
				activeSources = append(activeSources, source)
				break
			}
		}
	}
	return activeSources
}

func (sm *SourceManager) GetActiveSourceIds() []string {
	return sm.state.ActiveSourceIds
}

func (sm *SourceManager) GetSourceByID(id string) *KnowledgeSource {
	// First try exact UUID match
	for _, source := range sm.state.AllSources {
		if source.ID == id {
			return &source
		}
	}

	// Then try partial UUID match (for backward compatibility with 4-char prefixes)
	for _, source := range sm.state.AllSources {
		if strings.HasPrefix(source.ID, id) {
			return &source
		}
	}

	// Then try sequential number match
	if number, err := strconv.Atoi(id); err == nil {
		for _, source := range sm.state.AllSources {
			if source.Number == number {
				return &source
			}
		}
	}

	return nil
}

func (sm *SourceManager) GetSourceByNumber(number int) *KnowledgeSource {
	for _, source := range sm.state.AllSources {
		if source.Number == number {
			return &source
		}
	}
	return nil
}

func (sm *SourceManager) GetSourceIdByNumber(number int) string {
	for _, source := range sm.state.AllSources {
		if source.Number == number {
			return source.ID
		}
	}
	return ""
}

func (sm *SourceManager) GetAllSources() []KnowledgeSource {
	return sm.state.AllSources
}

func (sm *SourceManager) UpdateSourceContent(sourceId string, newContent string) bool {
	source := sm.GetSourceByID(sourceId)
	if source == nil || source.Type != "file" {
		return false
	}

	// Write the new content back to the file
	err := ioutil.WriteFile(source.Path, []byte(newContent), 0644)
	if err != nil {
		return false
	}

	// After updating the file, we should re-ingest it to update the chunks
	return true
}

func (sm *SourceManager) GetSourceContent(sourceId string) string {
	source := sm.GetSourceByID(sourceId)
	if source == nil || source.Type != "file" {
		return ""
	}

	content, err := ioutil.ReadFile(source.Path)
	if err != nil {
		return ""
	}

	return string(content)
}

func (sm *SourceManager) RemoveSource(sourceId string) bool {
	// Try to find the source by UUID, partial UUID, or sequential number
	var matchingSource *KnowledgeSource

	// First try exact UUID match
	for _, source := range sm.state.AllSources {
		if source.ID == sourceId {
			matchingSource = &source
			break
		}
	}

	// Then try partial UUID match
	if matchingSource == nil {
		for _, source := range sm.state.AllSources {
			if strings.HasPrefix(source.ID, sourceId) {
				matchingSource = &source
				break
			}
		}
	}

	// Then try sequential number match
	if matchingSource == nil {
		if number, err := strconv.Atoi(sourceId); err == nil {
			for _, source := range sm.state.AllSources {
				if source.Number == number {
					matchingSource = &source
					break
				}
			}
		}
	}

	if matchingSource == nil {
		return false
	}

	initialLength := len(sm.state.AllSources)

	// Remove from all sources
	var newAllSources []KnowledgeSource
	for _, source := range sm.state.AllSources {
		if source.ID != matchingSource.ID {
			newAllSources = append(newAllSources, source)
		}
	}
	sm.state.AllSources = newAllSources

	// Also remove from active sources if it was active
	var newActiveIds []string
	for _, id := range sm.state.ActiveSourceIds {
		if id != matchingSource.ID {
			newActiveIds = append(newActiveIds, id)
		}
	}
	sm.state.ActiveSourceIds = newActiveIds

	if len(sm.state.AllSources) < initialLength {
		sm.saveSources()
		return true
	}
	return false
}

// Helper function to generate a UUID
func generateUUID() string {
	bytes := make([]byte, 16)
	rand.Read(bytes)
	bytes[6] = (bytes[6] & 0x0f) | 0x40 // Version 4
	bytes[8] = (bytes[8] & 0x3f) | 0x80 // Variant 10
	return hex.EncodeToString(bytes)
}
